### PR TITLE
[EraVM] Add disassembler C-API.

### PR DESCRIPTION
**Pull request by PavelKopyl**
Example:
Default mode:
```
       add	128, r0, r3
	stm.h	64, r3
	and!	1, r2, r0
	jump.ne	30
	add	r1, r0, r2
	shr.s	96, r2, r2
	and	code[11], r2, r2
	sub.s!	4, r2, r0
	jump.lt	38
	ldp	r1, r3
	and	code[13], r3, r3
	sub.s!	code[14], r3, r0
	jump.ne	38
	sub.s!	68, r2, r0
	jump.lt	38
	ldvl	r2
	sub!	r2, r0, r0
	jump.ne	38
	addp.s	36, r1, r2
	ldp	r2, r2
	rol	code[15], r2, r3
	sub.s	1, r0, r4
	sub.s!	255, r2, r0
	add.le	r3, r0, r4
	addp.s	4, r1, r1
	ldp	r1, r1
	and	r4, r1, r1
	stm.h	128, r1
	add	code[16], r0, r1
	retl 41
	ldvl	r1
	sub!	r1, r0, r0
	jump.ne	38
	add	32, r0, r1
	stm.ah	256, r1
	stm.ah	288, r0
	add	code[12], r0, r1
	retl 41
	add	r0, r0, r1
	revl 42
	pncl	40
	retl 41
	revl 42
	<unknown> // zero fill for padding
11:
	.cell 4294967295
12:
	.cell 53919893334301279589334030174039261352344891250716429051063678533632
13:
	.cell -26959946667150639794667015087019630673637144422540572481103610249216
14:
	.cell 6775656105105473604099007350066834251830933750474818692318744137155055976448
15:
	.cell 18446744073709551614
16:
	.cell 2535301202817642044428229017600
	<unknown> // Next four octets are zero fill for padding
	<unknown>
	<unknown>
	<unknown>
	<unknown> // Next four octets are the hash. 
	<unknown>
	<unknown>
	<unknown>
```

Verbose mode:
```
 0: 00 00 00 80 03 00 00 39	add	128, r0, r3
       8: 00 00 00 40 00 30 04 3f	stm.h	64, r3
      10: 00 00 00 01 00 20 01 90	and!	1, r2, r0
      18: 00 00 00 1e 00 00 c1 3d	jump.ne	30
      20: 00 00 00 00 02 01 00 19	add	r1, r0, r2
      28: 00 00 00 60 02 20 02 70	shr.s	96, r2, r2
      30: 00 00 00 0b 02 20 01 97	and	code[11], r2, r2
      38: 00 00 00 04 00 20 00 8c	sub.s!	4, r2, r0
      40: 00 00 00 26 00 00 41 3d	jump.lt	38
      48: 00 00 00 00 03 01 04 3b	ldp	r1, r3
      50: 00 00 00 0d 03 30 01 97	and	code[13], r3, r3
      58: 00 00 00 0e 00 30 00 9c	sub.s!	code[14], r3, r0
      60: 00 00 00 26 00 00 c1 3d	jump.ne	38
      68: 00 00 00 44 00 20 00 8c	sub.s!	68, r2, r0
      70: 00 00 00 26 00 00 41 3d	jump.lt	38
      78: 00 00 00 00 02 00 04 16	ldvl	r2
      80: 00 00 00 00 00 02 00 4b	sub!	r2, r0, r0
      88: 00 00 00 26 00 00 c1 3d	jump.ne	38
      90: 00 00 00 24 02 10 03 70	addp.s	36, r1, r2
      98: 00 00 00 00 02 02 04 3b	ldp	r2, r2
      a0: 00 00 00 0f 03 20 02 df	rol	code[15], r2, r3
      a8: 00 00 00 01 04 00 00 8a	sub.s	1, r0, r4
      b0: 00 00 00 ff 00 20 00 8c	sub.s!	255, r2, r0
      b8: 00 00 00 00 04 03 a0 19	add.le	r3, r0, r4
      c0: 00 00 00 04 01 10 03 70	addp.s	4, r1, r1
      c8: 00 00 00 00 01 01 04 3b	ldp	r1, r1
      d0: 00 00 00 00 01 14 01 6f	and	r4, r1, r1
      d8: 00 00 00 80 00 10 04 3f	stm.h	128, r1
      e0: 00 00 00 10 01 00 00 41	add	code[16], r0, r1
      e8: 00 00 00 29 00 01 04 2e	retl 41
      f0: 00 00 00 00 01 00 04 16	ldvl	r1
      f8: 00 00 00 00 00 01 00 4b	sub!	r1, r0, r0
     100: 00 00 00 26 00 00 c1 3d	jump.ne	38
     108: 00 00 00 20 01 00 00 39	add	32, r0, r1
     110: 00 00 01 00 00 10 04 43	stm.ah	256, r1
     118: 00 00 01 20 00 00 04 43	stm.ah	288, r0
     120: 00 00 00 0c 01 00 00 41	add	code[12], r0, r1
     128: 00 00 00 29 00 01 04 2e	retl 41
     130: 00 00 00 00 01 00 00 19	add	r0, r0, r1
     138: 00 00 00 2a 00 01 04 30	revl 42
     140: 00 00 00 28 00 00 04 32	pncl	40
     148: 00 00 00 29 00 01 04 2e	retl 41
     150: 00 00 00 2a 00 01 04 30	revl 42
     158: 00 00 00 00 00 00 00 00	<unknown>
11:
	.cell 4294967295
12:
	.cell 53919893334301279589334030174039261352344891250716429051063678533632
13:
	.cell -26959946667150639794667015087019630673637144422540572481103610249216
14:
	.cell 6775656105105473604099007350066834251830933750474818692318744137155055976448
15:
	.cell 18446744073709551614
16:
	.cell 2535301202817642044428229017600
     220: 00 00 00 00 00 00 00 00	<unknown>
     228: 00 00 00 00 00 00 00 00	<unknown>
     230: 00 00 00 00 00 00 00 00	<unknown>
     238: 00 00 00 00 00 00 00 00	<unknown>
     240: 3a c2 25 16 8d f5 42 12	<unknown>
     248: a2 5c 1c 01 fd 35 be bf	<unknown>
     250: ea 40 8f da c2 e3 1d dd	<unknown>
     258: 6f 80 a4 bb f9 a5 f1 cb	<unknown>
``` 

# Code Review Checklist

## Requirements
- [ ] Have the requirements been met?
- [ ] Have stakeholder(s) approved the change?

## Implementation
- [ ] Does this code change accomplish what it is supposed to do?
- [ ] Can this solution be simplified?
- [ ] Does this change add unwanted compile-time or run-time dependencies?
- [ ] Could an additional framework, API, library, or service improve the solution?
- [ ] Could we reuse part of LLVM instead of implementing the patch or a part of it?
- [ ] Is the code at the right abstraction level?
- [ ] Is the code modular enough?
- [ ] Can a better solution be found in terms of maintainability, readability, performance, or security?
- [ ] Does similar functionality already exist in the codebase? If yes, why isn’t it reused?
- [ ] Are there any best practices, design patterns or language-specific patterns that could substantially improve this code? 

## Logic Errors and Bugs
- [ ] Can you think of any use case in which the
code does not behave as intended?
- [ ] Can you think of any inputs or external events
that could break the code?

## Error Handling and Logging
- [ ] Is error handling done the correct way?
- [ ] Should any logging or debugging information
be added or removed?
- [ ] Are error messages user-friendly?
- [ ] Are there enough log events and are they
written in a way that allows for easy
debugging?

## Maintainability
- [ ] Is the code easy to read?
- [ ] Is the code not repeated (DRY Principle)?
- [ ] Is the code method/class not too long?

## Dependencies
- [ ] Were updates to documentation, configuration, or readme files made as required by this change?
- [ ] Are there any potential impacts on other parts of the system or backward compatibility?

## Security
- [ ] Does the code introduce any security vulnerabilities?

## Performance
- [ ] Do you think this code change decreases
system performance?
- [ ] Do you see any potential to improve the
performance of the code significantly?

## Testing and Testability
- [ ] Is the code testable?
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
	- [ ] For changes to mutable state
- [ ] Do tests reasonably cover the code change (unit/integration/system tests)? 
	- [ ] Line Coverage
	- [ ] Region Coverage
	- [ ] Branch Coverage
- [ ] Are there some test cases, input or edge cases
that should be tested in addition?

## Readability
- [ ] Is the code easy to understand?
- [ ] Which parts were confusing to you and why?
- [ ] Can the readability of the code be improved by
smaller methods?
- [ ] Can the readability of the code be improved by
different function, method or variable names?
- [ ] Is the code located in the right
file/folder/package?
- [ ] Do you think certain methods should be
restructured to have a more intuitive control
flow?
- [ ] Is the data flow understandable?
- [ ] Are there redundant or outdated comments?
- [ ] Could some comments convey the message
better?
- [ ] Would more comments make the code more
understandable?
- [ ] Could some comments be removed by making the code itself more readable?
- [ ] Is there any commented-out code?
- [ ] Have you run a spelling and grammar checker?

## Documentation
- [ ] Is there sufficient documentation?
- [ ] Is the ReadMe.md file up to date?

## Best Practices
- [ ] Follow Single Responsibility principle?
- [ ] Are different errors handled correctly?
- [ ] Are errors and warnings logged?
- [ ] Magic values avoided?
- [ ] No unnecessary comments?
- [ ] Minimal nesting used?

## Experts' Opinion
- [ ] Do you think a specific expert, like a security
expert or a usability expert, should look over
the code before it can be accepted?
- [ ] Will this code change impact different teams, and should they review the change as well?

### CODE REVIEW

**Review by hedgar2017 — APPROVED** [review]

File: llvm/lib/MC/MCC/AssemblerC.cpp [hunk 1]
```diff
@@ -186,3 +193,118 @@ LLVMBool LLVMExceedsSizeLimitEraVM(LLVMMemoryBufferRef MemBuf,
...
+      uint64_t WordNum = 0;
+      to_integer<uint64_t>(Matches[1], WordNum, /*Base=*/10);
+      WordRefs.insert(WordNum);
+    }
+    // If we are at the word boundary and the word is being referenced,
+    // emit the numeric label and the .cell directive. Also set up the
+    // flag that we are collecting the cell parts.
+    if (uint64_t WordNum = PC / WordSize;
+        (PC % WordSize) == 0 && WordRefs.count(WordNum)) {
+      CollectCellBytes = true;
```
Comment (akiramenai) [inline llvm/lib/MC/MCC/AssemblerC.cpp]: Can't we perform all data-related handling in a single `if` instead of introducing `CollectCellBytes`?

File: llvm/lib/MC/MCC/AssemblerC.cpp [hunk 2]
```diff
@@ -186,3 +193,118 @@ LLVMBool LLVMExceedsSizeLimitEraVM(LLVMMemoryBufferRef MemBuf,
...
+    std::memcpy(InstrBytes, Bytes + PC, InstrSize);
+    size_t NumRead = LLVMDisasmInstruction(DCR, InstrBytes, InstrSize, /*PC=*/0,
+                                           OutString, OutStringSize);
+    // Check if the instruction contains a code reference. If so,
+    // extract the word number and add it to the WordRefs set.
+    SmallVector<StringRef, 2> Matches;
+    if (NumRead && CodeRegex.match(OutString, &Matches)) {
+      uint64_t WordNum = 0;
+      to_integer<uint64_t>(Matches[1], WordNum, /*Base=*/10);
+      WordRefs.insert(WordNum);
```
Comment (akiramenai) [inline llvm/lib/MC/MCC/AssemblerC.cpp]: If we allow data to be everywhere, shouldn't we first check that we don't disassemble already referenced constants?

**Review by akiramenai — CHANGES_REQUESTED** [review]
We could disassemble a constant and find that it 'references' something else in the `code` section. Then we mistakingly emit it as a `.cell`.

File: llvm/lib/MC/MCC/AssemblerC.cpp (hunk 1 above)
Comment (PavelKopyl) [reply llvm/lib/MC/MCC/AssemblerC.cpp]: Sure, will change this.

File: llvm/lib/MC/MCC/AssemblerC.cpp (hunk 2 above)
Comment (PavelKopyl) [reply llvm/lib/MC/MCC/AssemblerC.cpp]: You are right, but I implemented this relying on the current binary layout, where all the `.rodata` always comes after the instructions.  Do you mean we have to support arbitrary binary layout in disassembler?

**Comment by PavelKopyl** [comment]
> We could disassemble a constant and find that it 'references' something else in the `code` section. Then we mistakingly emit it as a `.cell`.

Could you please, clarify that? We identify the location of a constant by analyzing `code[word_num]` arguments of instructions. If a such instruction exists, then at the location `word_num`, the we definitely have to emit `.cell `with the constant.

**Review by akiramenai — APPROVED** [review]
LGTM, thanks
Please squash.
